docs/content/v0.20/architecture.md

```diff
@@ -1,36 +1,433 @@
+    events where the object content is not modified) and then creates a
+    key to store a reference to the object in the work queue. The key
+    contains the GVK key used to retrieve a reference to the informer
+    and lister for that kind of object, and a namespace + name key to
+    retrieve the actual object. Storing the key in the work queue is a
+    common pattern in client-go as the object may have changed in the
+    cache (which is kept updated by the reflector) by the time a worker
+    gets a copy from the work queue. Workers should always receive the
+    key and use it to retrieve the object from the cache.
+
+3.  A worker pulls a key from the work queue, and then does the
+    following processing:
+
+    -  Uses the GVK key to get the reference to the lister for the
+        object
+    -  Gets the object from the lister cache using the NamespacedName of 
+       the object.
+    -  Gets the lister for Placement objects and list all placements
+    -  Iterates on all placements, and for each placement:
+        - evaluates whether the object matches the downsync selection 
+          criteria in the Placement.
+        - If there is a match, list ManagedClusters and
+          find the matching clusters using the label selector expression
+          for clusters.
+        - If there are matching clusters, add the names of the cluster
+          to a hashmap setting the name of the cluster as a key. Cluster
+          groups from different placements are merged together.
+        - If any of the matched Placements has ` WantSingletonReportedState`
+          set to true, clusters are sorted in alphanumerical order and
+          only the first cluster is selected for delivery.  
+    -  If there are no matching clusters, the worker returns without
+        actions and is ready to process other events from the queue.
+    -  If there are matching clusters:
+        - Wraps the object into a ManifestWork
+        - Adds a label for each matched Placement to the ManifestWork that 
+          is used to track the placement that caused the object to be 
+          delivered to one or more clusters. The label contains both the 
+          placement name (note that placement is cluster-scoped) and the WDS name. 
+          This way, when a placement is deleted or updated it is possible to 
+          locate the associated ManifestWorks for deletion or label removal 
+          (if more than one label is present, as other placements may “own” the object).
+        - For each cluster:
+          -  Sets the manifest namespace == name of the cluster
+          -  Uses the client connected to the ITS to do a server-side
+              apply of the manifest into the namespace.
+        - Worker returns and is ready to pick other keys from the queue.
+
+There are other event flows, based on the object GVK and type of event. 
+Error conditions cause the re-enqueing of keys, resulting in retries.
```

Comment:
Presumably not _all_ error conditions: the ones that can not be cured by retry should not be retried.

Reply:
Changed into "Error conditions may cause..." to account for that.

Reply:
My point is that not all error conditions do cause retries (if I understand correctly), only a subset. The error conditions that clearly will persist are not retried, right?

Reply:
Correct.